docs/architecture-decision-records/2023-03-integration-with-station.md

```diff
@@ -0,0 +1,260 @@
+  $HOME/Library/Caches/Filecoin\ Station/zinnia-modules/ping.js \
+  # Since we don't support ES Modules yet, a more complex module may want
+  # to bundle multiple source code files into a single file for distribution
+  # In such case, a module can be a directory with nested subdirectories too.
+  $HOME/Library/Caches/Filecoin\ Station/zinnia-modules/retrieval-checker/dist/index.js
+```
+
+### Communication with Station (Core, Desktop):
+
+As explained above, I propose to use environment variables to pass configuration from the Station to
+Zinnia.
+
+For communicating information from Zinnia back to the Station, I propose that Zinnia prints
+newline-delimited JSON entries to `stdout` and uses `stderr` for general/debug logging. This format
+is easy to parse from Node.js code powering both Station variants.
+
+Example messages:
+
+- **Activity log - error**
+
+  `{"type": "activity:error", "module": "saturn", "message": "Cannot connect to the orchestrator."x }`
+
+  _Note: `"module": "saturn"` describes which module emitted the log._
+
+- **Activity log - info**
+
+  `{"type": "activity:info", "module": null, "message": "Zinnia is starting up..."}`
+
+  _Note: `"module": null` means the message comes from Zinnia runtime._
+
+- **Number of jobs completed:**
+
+  `{"type": "jobs-completed", "total": 123 }`
+
+  _Note: This message is emitted periodically, e.g. every 200ms._
+
+  In the future, we can easily extend this line to include per-module stats too:
+
+  `{"type": "jobs-completed", "total": 123, modules: {"saturn": 100, "retrieval-checker": 23}}`
+
+### Module identifiers
+
+We need each module to have a unique identifier (a name) that we can use in the messages above. This
+id must remain unchanged across module version upgrades.
+
+For the initial version, these ids will be hard-coded human-readable names like `saturn-l2`.
+
+In the future, when we move towards untrusted modules deployed in a decentralized manner, we will
+need to find a different way how to derive these unique module ids. That's out of scope of the
+current work though. Adding new id types should be easy as long as our architecture supports
+arbitrary string ids.
+
+### Deploying and upgrading modules
+
+The initial version will not implement any upgrade mechanisms for modules. Both `zinniad` and all
+module sources will be bundled inside the Station. When a new module is added or an existing module
+is upgraded to a new version, we will publish a new version of the Station.
+
+### Zinnia API for module builders
+
+_(This is a strawman proposal, subject to changes during implementation.)_
+
+```ts
+namespace Zinnia {
+  // omitted: existing APIs like `peerId`
+
+  /** Get the wallet address, this value is typically provided by the Station. */
+  walletAddress: String;
+
+  /** Report activities to the Station */
+  log: {
+    /** Report an informative status update, e.g. "Connecting to the network." */
+    info(message: string);
```

Comment:
Definitely not!

Messages printed by `Zinnia.log.info` are shown in the Station UI. Modules should use this API sparingly to not clutter the UI. We may want to enforce that somehow in the future.

Messages printed by Console APIs like `console.log` are considered "debug" logs and are not visible to the Station user by default.

IMO, we should also extend the current Station feature for downloading module logs to download logs from Zinnia modules.

Reply:
Would it help to rename this API to something like `Zinnia.activityLog.info` to clarify that these messages are not regular logs but have special meaning?

Reply:
Gotcha, I didn't know this was about the activity log. Yeah, `Zinnia.activityLog.info`, or even `Zinnia.activity.info` would be clearer to me

> IMO, we should also extend the current Station feature for downloading module logs to download logs from Zinnia modules.

Yes! The current feature only downloads logs from one module, because we only had one. It needs to be extended once we have multiple.

Reply:
Cool, I am going to use `Zinnia.activity.info`.